lcls-twincat-motion/Library/POUs/Motion/PMPS/FB_MiscStatesErrorFFO.TcPOU

```diff
@@ -0,0 +1,125 @@
+      This solves an issue where ffBeamParamsOK might solve its own problem and creating a blinking fast fault issue.
+
+    The inputs are mostly outputs of non-pmps function blocks and heavily-reused info like state details.
+*)
+VAR_IN_OUT
+    // The arbiter to request beam states from
+    fbArbiter: FB_Arbiter;
+    // The fast fault output to fault to
+    fbFFHWO: FB_HardwareFFOutput;
+END_VAR
+VAR_INPUT
+    // A name to link to these fast faults
+    sDeviceName: STRING;
+    // Current requested beam details: either a known state or the transition beam
+    stCurrentBeamReq: ST_BeamParams;
+    // TRUE if we're at a known state (doesn't matter which)
+    bKnownState: BOOL;
+    // Lookup ID of the transition beam
+    nTransitionID: DWORD;
+END_VAR
+VAR_OUTPUT
+END_VAR
+VAR CONSTANT
+    // Number of consecutive trips before we debounce
+    nMaxTrips: UINT := 5;
+    // Decrease trip count by 1 after this much time has passed
+    tTripReset: TIME := T#1s;
+END_VAR
+VAR
+    // If the beam parameters are wrong, it is a fault! This encompasses all unknown arbiter-related errors.
+    ffBeamParamsOk: FB_FastFault;
+    // If we asked for zero rate (NC or SC) then we can cut the beam early. This is somewhat redundant.
+    ffZeroRate: FB_FastFault;
+    // Trip the beam for unknown state
+    ffUnknown: FB_FastFault;
+    // Trip the beam (no autoreset) if any of the previous FBs are tripping/resetting too quickly.
+    ffDebounce: FB_FastFault;
+
+    // Number of consecutive trips so far
+    nTripCount: UINT;
+    // Increase by 1 whenever there is a fault (rising edge)
+    ftTripCount: F_TRIG;
+    // Decrease trip count by 1 each timeout
+    tonTripCount: TON;
+    // TRUE on only the first cycle
+    bFirstCycle: BOOL := TRUE;
+END_VAR
+]]></Declaration>
+    <Implementation>
+      <ST><![CDATA[
+ffBeamParamsOk(
+    i_xOK:=F_SafeBPCompare(PMPS_GVL.stCurrentBeamParameters, stCurrentBeamReq),
+    i_xAutoReset:=TRUE,
+    i_DevName:=sDeviceName,
+    i_Desc:='Beam parameter mismatch',
+    i_TypeCode:=16#1000,
+    io_fbFFHWO:=fbFFHWO);
+
+CASE PMPS_GVL.stCurrentBeamParameters.nMachineMode OF
+    // NC mode
+    1: ffZeroRate.i_xOK := stCurrentBeamReq.nRate > 0;
+    // SC mode
+    2: ffZeroRate.i_xOK := stCurrentBeamReq.nBCRange > 0;
+ELSE
+    // Ambiguous or new mode
+    ffZeroRate.i_xOK := stCurrentBeamReq.nRate > 0 AND stCurrentBeamReq.nBCRange > 0;
+END_CASE
+
+ffZeroRate(
+    i_xAutoReset := TRUE,
+    i_DevName := sDeviceName,
+    i_Desc := 'Device requesting zero rate',
+    i_TypeCode := 16#1001,
+    io_fbFFHWO := fbFFHWO,
+);
+
+ffUnknown(
+    i_xOK := bknownState OR fbArbiter.CheckRequestInPool(nReqID:=nTransitionID),
+    i_xAutoReset := TRUE,
+    i_DevName := sDeviceName,
+    i_Desc := 'Unknown position between moves',
+    i_TypeCode := 16#1005,
+    io_fbFFHWO := fbFFHWO,
+);
+
+ftTripCount(CLK:=ffBeamParamsOk.i_xOK);
```

Comment:
You're right, but I think the correct thing to do is to change the docstring. The beam params ok is the only one that needs this sort of protection to avoid the situation where the act of doing a trip clears the trip condition via removing the beam, making the beam params ok.